torchtune/models/qwen2_5/_tokenizer.py

```diff
@@ -0,0 +1,232 @@
+        bpe_cache_size: int = DEFAULT_QWEN2_TOKENIZER_BPE_CACHE_SIZE,
+    ):
+        super().__init__(
+            path=path,
+            merges_file=merges_file,
+            special_tokens=special_tokens,
+            max_seq_len=max_seq_len,
+            prompt_template=prompt_template,
+            errors=errors,
+            unk_token=unk_token,
+            bos_token=bos_token,
+            eos_token=eos_token,
+            pad_token=pad_token,
+            bpe_cache_size=bpe_cache_size,
+        )
+
+    def tokenize_messages(
+        self,
+        messages: List[Message],
+        *,
+        add_eos: bool = True,
+    ) -> Tuple[List[int], List[bool]]:
+        """
+        Given a list of messages, return a list of tokens for the concatenated
+        and formatted messages.
+
+        Args:
+            messages (List[Message]): The message list to tokenize.
+            add_eos (bool): Wether to add the tokenizer's eos_id at the end of the
+                sequence of messages. Default is True.
+
+        Returns:
+            Tuple[List[int], List[bool]]: The list of token ids and the list of masks.
+
+        Raises:
+            RuntimeError: If a message contains non-text content
+        """
+        assert not isinstance(
+            self.prompt_template, ChatMLTemplate
+        ), "Using ChatMLTemplate with tokenize_messages will result in multiple <|im_*|> tokens wrapping each message."
+        templated_messages = (
+            self.prompt_template(messages)
+            if self.prompt_template is not None
+            else messages
+        )
+
+        tokenized_messages = []
+        mask = []
+        for i, message in enumerate(templated_messages):
+            tokens = []
+
+            # message header
+            if message.role == "ipython":
+                if i == 0 or templated_messages[i - 1].role != "ipython":
+                    self._add_message_start_tokens(tokens, "user")
+                    tokens.extend(
+                        self.encode("<tool_response>\n", add_bos=False, add_eos=False)
+                    )
+                else:
+                    tokens.extend(
+                        self.encode("\n<tool_response>\n", add_bos=False, add_eos=False)
+                    )
+            else:
+                self._add_message_start_tokens(tokens, message.role)
+                if message.role == "assistant" and message.ipython:
+                    tokens.extend(
+                        self.encode("<tool_call>\n", add_bos=False, add_eos=False)
+                    )
+
+            # message content
+            for item in message.content:
+                if item["type"] == "text":
+                    tokens.extend(
+                        self.encode(
+                            item["content"],
+                            add_bos=False,
+                            add_eos=False,
+                        )
+                    )
+                else:
```

Comment:
does Qwen2.5 not support images?

Reply:
oops yeah it does, I'll look into how we should encode images